### PR TITLE
Parallel requests to graphite, added timeouts to the operation of getting unhealthy subscriptions list (#1323)

### PR DESCRIPTION
**Pull request by Kornelos**
Added parallel requests to graphite with timeout configurable for the whole operation #1323

### CODE REVIEW

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/SubscriptionHealthCheckProperties.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package pl.allegro.tech.hermes.management.config;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("subscription-health-check")
+public class SubscriptionHealthCheckProperties {
+    private long subscriptionHealthCheckTimeoutMillis = 8000;
+    private int subscriptionServiceThreads = 16;
```
Comment (jewertow) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/SubscriptionHealthCheckProperties.java:8]: For now the configuration file will look like this:
```
subscription-health-check:
  subscription-health-check-timeout-millis: 8000
  subscription-service-threads: 16
```
It seems a bit verbose. What do you think about simpler names like these:
```
subscription-health-check:
  timeout-millis: 8000
  threads: 16
```

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/SubscriptionServiceConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                topicService,
+                metricsRepository,
+                subscriptionHealthChecker,
+                logRepository,
+                subscriptionValidator,
+                auditor,
+                multiDcExecutor,
+                multiDCAwareService,
+                repositoryManager,
+                Executors.newFixedThreadPool(healthCheckProperties.getSubscriptionServiceThreads()),
```
Comment (jewertow) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/SubscriptionServiceConfiguration.java:50]: As far as I remember, it is preferred in the project to create threads with meaningful names using the `ThreadFactoryBuilder` like [here](https://github.com/allegro/hermes/blob/master/hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/health/HealthCheckScheduler.java#L29).

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/subscription/SubscriptionService.java [hunk 3]
```diff
@@ -274,20 +282,63 @@ public SubscriptionHealth getSubscriptionHealth(TopicName topicName, String subs
...
+                    .thenComposeAsync(ownerSubscriptions ->
+                                    getUnhealthyList(respectMonitoringSeverity, subscriptionNames, qualifiedTopicNames, ownerSubscriptions),
+                            subscriptionHealthCheckExecutorService).
+                            get(subscriptionHealthCheckTimeoutMillis, TimeUnit.MILLISECONDS);
+
+        } catch (TimeoutException e) {
+            throw new UnhealthySubscriptionGetException("Fetching unhealthy subscriptions timed out.");
+        } catch (Exception e) {
+            throw new UnhealthySubscriptionGetException("Fetching unhealthy subscriptions failed.", e);
+        }
```
Comment (piotrrzysko) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/subscription/SubscriptionService.java:314]: What do you think about moving more common code to `getUnhealthyList`?: 
```java

    public List<UnhealthySubscription> getAllUnhealthy(boolean respectMonitoringSeverity,
                                                       List<String> subscriptionNames,
                                                       List<String> qualifiedTopicNames) {
        return getUnhealthyList(subscriptionOwnerCache.getAll(), respectMonitoringSeverity, subscriptionNames, qualifiedTopicNames);
    }

    public List<UnhealthySubscription> getUnhealthyForOwner(OwnerId ownerId,
                                                            boolean respectMonitoringSeverity,
                                                            List<String> subscriptionNames,
                                                            List<String> qualifiedTopicNames) {
        return getUnhealthyList(subscriptionOwnerCache.get(ownerId), respectMonitoringSeverity, subscriptionNames, qualifiedTopicNames);
    }

    private List<UnhealthySubscription> getUnhealthyList(Collection<SubscriptionName> ownerSubscriptionNames,
                                                         boolean respectMonitoringSeverity,
                                                         List<String> subscriptionNames,
                                                         List<String> qualifiedTopicNames) {
        try {
            return getSubscriptionDetails(ownerSubscriptionNames)
                    .thenComposeAsync(ownerSubscriptions -> {
                                List<CompletableFuture<UnhealthySubscription>> unhealthySubscriptions = filterSubscriptions(
                                        ownerSubscriptions,
                                        respectMonitoringSeverity,
                                        subscriptionNames,
                                        qualifiedTopicNames
                                );
                                return CompletableFuture.allOf(unhealthySubscriptions.toArray(new CompletableFuture[0])).thenApply(
                                        v -> unhealthySubscriptions.stream()
                                                .map(CompletableFuture::join)
                                                .filter(Objects::nonNull)
                                                .collect(Collectors.toList())
                                );
                            },
                            subscriptionHealthCheckExecutorService
                    )
                    .get(subscriptionHealthCheckTimeoutMillis, TimeUnit.MILLISECONDS);
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.FAILED_DEPENDENCY);
        }
    }
```

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/UnhealthySubscriptionsGetProperties.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package pl.allegro.tech.hermes.management.config;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("unhealthy-subscription-get")
```
Comment (piotrrzysko) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/UnhealthySubscriptionsGetProperties.java:5]: There is a class with properties related to this mechanism: `pl.allegro.tech.hermes.management.config.HealthProperties`. Unfortunately its name is a little bit confusing - somebody may think that it is an application’s health check configuration. I suggest to change current structure to something like this:
```yaml
subscription-health-check: 
    maxLagInSeconds: 600
    maxOtherErrorsRatio: 0.5
    ...
    timeoutMillis: 8000
    threadPoolSize: 16
```

Comment (Kornelos) [reply hermes-management/src/main/java/pl/allegro/tech/hermes/management/config/UnhealthySubscriptionsGetProperties.java]: I've changed its name to `SubscriptionHealthProperties ` and added timoeut and threadPool properties there

**Review by piotrrzysko — APPROVED** [review]

**Review by druminski — APPROVED** [review]

**Pull request merged**